File: src/main/java/com/wanted/preonboarding/backend20/global/auth/JwtTokenProvider.java
package com.wanted.preonboarding.backend20.global.auth;

import io.jsonwebtoken.*;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.security.Key;
import java.util.Date;
import java.util.stream.Collectors;

@Component
public class JwtTokenProvider {
    private final Key key;
    private final UserDetailsServiceImpl userDetailsService;

    private static final long ACCESS_TOKEN_EXPIRE_TIME = 30 * 60 * 1000L;

    public JwtTokenProvider(@Value("${jwt.secret-key}") String secretKey, UserDetailsServiceImpl userDetailsService) {
        byte[] keyBytes = Decoders.BASE64.decode(secretKey);
        this.key = Keys.hmacShaKeyFor(keyBytes);
        this.userDetailsService = userDetailsService;
    }

    public String generateToken(Authentication authentication) {
        String authorities = authentication.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.joining(","));

        Date now = new Date();
        Date expireDate = new Date(now.getTime() + ACCESS_TOKEN_EXPIRE_TIME);

        return Jwts.builder()
                .setSubject(authentication.getName())
                .claim("authorities", authorities)
                .setExpiration(expireDate)
                .signWith(key)
                .compact();
    }
    
    public boolean validateToken(String token) throws JwtException {
        try {
            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
            return true;
        } catch (SecurityException | MalformedJwtException e) {
            throw new JwtException("잘못된 타입의 토큰");
        } catch (ExpiredJwtException e) {
            throw new JwtException("만료된 토큰");
        } catch (UnsupportedJwtException e) {
            throw  new JwtException("지원되지 않는 토큰");
        } catch (IllegalArgumentException e) {
            throw new JwtException("토큰이 존재하지 않음");
        }
    }

    public Authentication getAuthentication(String accessToken) {
        Claims claims = Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(accessToken).getBody();
        UserDetails userDetails = userDetailsService.loadUserByUsername(claims.getSubject());
        return new UsernamePasswordAuthenticationToken(userDetails, accessToken, userDetails.getAuthorities());
    }

}

File: src/main/kotlin/wanted/market/api/common/jwt/JwtTokenProvider.kt
package wanted.market.api.common.jwt

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import jakarta.annotation.PostConstruct
import jakarta.servlet.http.HttpServletRequest
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import wanted.market.api.common.security.CustomMemberDetailService
import java.util.*

@Component
class JwtTokenProvider(
    private val customMemberDetailService: CustomMemberDetailService,
    @Value("\${jwt.secret}")
    private var secretKey: String? = null
) {

    @PostConstruct
    protected fun init() {
        secretKey = Base64.getEncoder().encodeToString(secretKey!!.toByteArray())
    }

    fun resolveToken(request: HttpServletRequest): String {
        return request.getHeader("Authorization")
    }

    fun decode(token: String?): Claims {
        return Jwts.parser()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .getBody()
    }

    // Jwt 토큰으로 인증 정보를 조회
    fun getAuthentication(token: String): Authentication {
        val claims = decode(token.substring("Bearer ".length))
        val userDetails = customMemberDetailService.loadUserByUsername(claims["memberId"].toString())
        return UsernamePasswordAuthenticationToken(userDetails, "", userDetails.authorities)
    }

    // Jwt 토큰의 유효성 + 만료일자 확인
    fun validateToken(jwtToken: String?): Boolean {
        val claims = Jwts.parser()
            .setSigningKey(secretKey)
            .build()
            .parseSignedClaims(jwtToken!!.substring("Bearer ".length))
        return !claims.getBody().expiration.before(Date()) // 만료시간이 현재시간보다 전인지 확인
    }
}
